File: SAMPLv0.1/SAMMcore/Components/Multipole.py
# classdef Multipole < handle
#     % Multipole class
#     %
#     % Properties:
#     %   name
#     %   angle
#     %   curvature
#     %   field
#     %   aperture
#     %
#     % Methods:
#     %   Track
#     %   TrackSpin
#     %   GetBField
#
#     properties
#         name      = ''; % string
#         angle     = 0;  % bending angle of reference trajectory, in radians
#         curvature = 0;  % reciprocal of radius of curvature of reference trajectory, in 1/metres (for applying weak focusing)
#         field     = []; % Nx2 array of N field components [index normal-component+i*skew-component]
#         aperture  = []; % 1x2 array of elliptical aperture half-axes, in metres
#     end % properties
#
#     properties (Constant=true)
#         length   = 0;
#     end % properties (constant)
#
#     methods
#
#
#         function beam = Track(multipole,beam)
#             % beam2 = Multipole.Track(beam1)
#             % Applies the transfer map for a multipole to the particles
#             % in beam1.
#
#             [x0, px0, y0, py0, ct0, dp0] = beam.GetParticles();
#
#             dpx = -multipole.angle * (1 + dp0 - multipole.curvature * x0);
#
#             % Apply a multipole 'kick'
#             xi0 = x0 + 1i*y0;
#
#             for indx = 1:size(multipole.field,1)
#                 n   = multipole.field(indx,1);
#                 dpx = dpx + multipole.field(indx,2) * (xi0.^n) / beam.rigidity / factorial(n);
#             end
#
#             px1 = px0 - real(dpx);
#             py1 = py0 + imag(dpx);
#             ct1 = ct0 - multipole.angle * x0;
#
#             beam.SetParticles(x0, px1, y0, py1, ct1, dp0);
#
#         end % function Track
#
#
#         function TrackLibrary(multipole,trackingMethod,libroutine)
#
#             multipole.field       = sortrows(multipole.field,1);
#
#             multipole1.angle      = multipole.angle;
#             multipole1.curvature  = multipole.curvature;
#             multipole1.ncpts      = size(multipole.field,1);
#             multipole1.fieldindex = libpointer('doublePtr',real(multipole.field(:,1)));
#             multipole1.bnL        = libpointer('doublePtr',real(multipole.field(:,2)));
#             multipole1.anL        = libpointer('doublePtr',imag(multipole.field(:,2)));
#             if(~isempty(multipole.aperture))
#                multipole1.apertureX = multipole.aperture(1);
#                multipole1.apertureY = multipole.aperture(2);
#             end
#
#             calllib(trackingMethod,[libroutine 'Multipole'],multipole1);
#
#         end % function TrackLibrary
#
#
#         function [bx, by, bz] = GetBField(multipole,beam)
#             % [bx, by, bz] = Multipole.GetBField(beam)
#             % Returns the *integrated* magnetic field (in tesla metres) at
#             % the locations of the particles in the beam.
#
#             [x, ~, y] = beam.GetParticles();
#
#             xi = x + 1i * y;
#             b  = 0;
#
#             for indx = 1:size(multipole.field,1)
#                 n = multipole.field(indx,1);
#                 b = b + multipole.field(indx,2) * xi.^n / factorial(n);
#             end
#
#             bx = real(b);
#             by = imag(b);
#             bz = 0;
#
#         end % function GetBField
#
#
#         function beam = TrackSpin(multipole,beam)
#             % beam2 = Multipole.Trackspin(beam1)
#             % Tracks particle spins through a Multipole.
#
#             [bxL, byL, bzL] = multipole.GetBField(beam);
#
#             nomlength = 1e-9;
#
#             Utilities.SpinRotation(beam, bxL/nomlength, byL/nomlength, bzL/nomlength, nomlength);
#
#             % Account for the rotation of the local coordinate system
#
#             if multipole.angle ~= 0
#
#                 [theta1, phi1] = beam.GetSpins();
#
#                 polsnx2    = sin(theta1).*cos(phi1);
#                 polsny2    = sin(theta1).*sin(phi1);
#                 polsnz2    = cos(theta1);
#
#                 thetaprime = acos(polsny2);
#                 phiprime   = atan2(polsnx2,polsnz2) + multipole.angle;
#
#                 polsnx3    = sin(thetaprime).*sin(phiprime);
#                 polsny3    = cos(thetaprime);
#                 polsnz3    = sin(thetaprime).*cos(phiprime);
#
#                 phi2       = atan2(polsny3,polsnx3);
#                 theta2     = acos(polsnz3);
#
#                 beam.SetSpins(theta2, phi2);
#
#             end
#
#         end % function TrackSpin
#
#
#     end % methods
#
# end % classdef Multipole

File: SAMM3.0/TrackingLibraries/CParticleTracking/Multipole.c

#include <math.h>
#include <stdlib.h>

#include "CParticleTracking.h"
#include "AppendComponent.h"
#include "GlobalVariables.h"

#define _apertureX 0
#define _apertureY 1
#define _bendangle 2
#define _curvature 3
#define _ncpts 4
#define _findex 5


// int factorial(int n)
// {
//	return (n==1 || n==0) ? 1 : factorial(n - 1) * n;
// }


void TrackMultipole_(double* parameters)
{
	const double ax        = parameters[_apertureX];
	const double ay        = parameters[_apertureY];

	const double bendangle = parameters[_bendangle];
	const double curvature = parameters[_curvature];

	const int    ncpts     = (int)parameters[_ncpts];

	double*      findex    = &parameters[_findex];
	double*      anL       = &parameters[_findex + ncpts];
	double*      bnL       = &parameters[_findex + ncpts*2];

	const double brho      = beamParams->rigidity;

	int          n;


	for(n=0; n<beamParams->nparticles; n++)
	{
		if(distance[n*2 + 1])
		{

			double x0   = beamPS[n*6];
			double px0  = beamPS[n*6 + 1];
			double y0   = beamPS[n*6 + 2];
			double py0  = beamPS[n*6 + 3];
			double ct0  = beamPS[n*6 + 4];
			double dp0  = beamPS[n*6 + 5];

			double BxL  = 0;
			double ByL  = 0;

			double psi  = atan2(y0,x0);
			double r    = sqrt(x0*x0 + y0*y0);

			int    indx = 0;
			int    ni   = 1;

			double pr   = 1;

			for(indx=0; indx<ncpts; indx++)
			{
				int nn    = (int) findex[indx];

				double cs = cos(nn*psi);
				double sn = sin(nn*psi);

				//double pr = pow(r,nn) / factorial(nn);
				for( ; ni<=nn; ni++)
					pr = pr * r / ni;

				ByL      += pr*(bnL[indx]*cs - anL[indx]*sn);
				BxL      += pr*(bnL[indx]*sn + anL[indx]*cs);
			}

			px0 += bendangle*(1 + dp0 - curvature*x0) - ByL/brho;
			py0 += BxL/brho;
			ct0 -= bendangle*x0;

			beamPS[n*6 + 1] = px0;
			beamPS[n*6 + 3] = py0;
			beamPS[n*6 + 4] = ct0;


			if(beamParams->g != 0) {
				double theta = beamSpins[n*2];
				double phi   = beamSpins[n*2 + 1];
				double L     = 1e-9;
				double BzL   = 0;

				RotateSpins(BxL/L, ByL/L, BzL/L, L, px0, py0, dp0, &theta, &phi);

				beamSpins[n*2] = theta;
				beamSpins[n*2 + 1] = phi;
			}


			if( (ax>0) & (ay>0) )
				if( (x0/ax)*(x0/ax) + (y0/ay)*(y0/ay) > 1)
					distance[n*2 + 1] = 0;

		}
	}

}


void TrackMultipole(MultipoleParameters_t multipole)
{
	int ncpts  = (int) multipole.ncpts;

	int n;

	double* parameters = (double*)calloc(5+3*ncpts,sizeof(double));

	parameters[_apertureX] = multipole.apertureX;
	parameters[_apertureY] = multipole.apertureY;

	parameters[_bendangle] = multipole.angle;
	parameters[_curvature] = multipole.curvature;

	parameters[_ncpts]     = (double)ncpts;

	for(n=0; n<ncpts; n++)
	{
		parameters[_findex + n]            = (double)multipole.fieldindex[n];
		parameters[_findex + n + ncpts]    = multipole.anL[n];
		parameters[_findex + n + ncpts*2]  = multipole.bnL[n];
	}

	TrackMultipole_(parameters);

	free(parameters);
}


void CopyMultipole(MultipoleParameters_t multipole)
{
	int ncpts  = (int) multipole.ncpts;

	int n;

	double* parameters = (double*)calloc(5+3*ncpts,sizeof(double));

	parameters[_apertureX] = multipole.apertureX;
	parameters[_apertureY] = multipole.apertureY;

	parameters[_bendangle] = multipole.angle;
	parameters[_curvature] = multipole.curvature;

	parameters[_ncpts]     = (double)ncpts;

	for(n=0; n<ncpts; n++)
	{
		parameters[_findex + n]            = (double)multipole.fieldindex[n];
		parameters[_findex + n + ncpts]    = multipole.anL[n];
		parameters[_findex + n + ncpts*2]  = multipole.bnL[n];
	}

	AppendComponent(TrackMultipole_, parameters);
}
